File: solutions/0046-permutations/permutations.java
// Given a collection of distinct integers, return all possible permutations.
//
// Example:
//
//
// Input: [1,2,3]
// Output:
// [
//   [1,2,3],
//   [1,3,2],
//   [2,1,3],
//   [2,3,1],
//   [3,1,2],
//   [3,2,1]
// ]
//
//


class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> all = new ArrayList<>();
        upset(nums, 0, all);
        return all;
    }

    private void upset(int[] nums, int begin, List<List<Integer>> all) {
        if(begin == nums.length) {
            ArrayList<Integer> temp = new ArrayList<Integer>(); 
            for(int i = 0; i < nums.length; i++) {
                temp.add(nums[i]);
            }
            all.add(new ArrayList<Integer>(temp));
            return;
        }
        
        for(int i = begin; i < nums.length; i++) {
            swap(nums, i, begin);
            upset(nums, begin + 1, all);
            swap(nums, i, begin);
        }

    }

    private void swap(int[] nums, int i, int begin) {
        int temp = nums[i];
        nums[i] = nums[begin];
        nums[begin] = temp;
    }
}

File: solutions/0046-permutations/permutations.cpp
// Given a collection of distinct integers, return all possible permutations.
//
// Example:
//
//
// Input: [1,2,3]
// Output:
// [
//   [1,2,3],
//   [1,3,2],
//   [2,1,3],
//   [2,3,1],
//   [3,1,2],
//   [3,2,1]
// ]
//
//


class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int> > ans;
        vector<int> combination;
        
        helper(nums, ans, combination);
        return ans;
    }
    
    void helper(const vector<int> nums, vector<vector<int> >& ans, vector<int> combination) {
        if(combination.size() == nums.size()) {
            ans.emplace_back(combination);
            return ;
        }
        
        for(int i=0; i<nums.size(); ++i) {
            if(find(combination.begin(), combination.end(), nums[i]) == combination.end()) {
                combination.emplace_back(nums[i]);
                helper(nums, ans, combination);
                combination.pop_back();
            }
        }
    } 
};
